### ISSUE TITLE
UTF-8 encoded mails are not sent

### ISSUE THREAD

**Issue by viktordick** [zopefoundation/zope.sendmail#27]
According to the documentation in `smptlib.sendmail`,

        msg may be a string containing characters in the ASCII range, or a byte
        string.  A string is encoded to bytes using the ascii codec, and lone
        \\r and \\n characters are converted to \\r\\n characters.

Since switching to Zope 4, we are unable to send mails that have a UTF-8 encoded body - Even if we transform it to bytes ourselves, `Products.MailHost` converts it back to strings before sending it to `zope.sendmail` and smptlib fails to encode the passed message using ASCII during the second phase of the two phase commit. Our quick fix was to change `zope.sendmail` to encode the message using `utf-8` before passing it to `smtplib.sendmail`.

A really clean way would probably be to inspect the headers (which should be ASCII) and check for a `Content-Type` and `Content-Transfer-Encoding`. Maybe it is even possible to have a multipart message with different encodings (not sure). But since UTF-8 is such a popular choice, we used that for now.

I am not sure if the fix should happen in `zope.sendmail` or somewhere further up like `Products.MailHost`. If it should be `zope.sendmail`, I could try to supply a PR. Otherwise I would need to be pointed to the right direction.

**Comment by d-maurer** [comment]
Viktor Dick wrote at 2020-7-14 00:01 -0700:
>According to the documentation in `smptlib.sendmail`,
>        msg may be a string containing characters in the ASCII range, or a byte
>        string.  A string is encoded to bytes using the ascii codec, and lone
>        \\r and \\n characters are converted to \\r\\n characters.

I read this as: the message must be ASCII only.
This is not surprising as it is a major requirement for low level mail
messages. If the orignal (high level) message contains non ASCII, it must be
encoded to ASCII (typically via "quoted-printable" or "base64").
Python's `email` package contains logic to do this.

**Comment by viktordick** [comment]
Thanks for the reply.

I actually read this as: Either ensure that you handled the encoding yourself (passing bytes) or only use ASCII. `stmplib` will not try to guess the encoding.

I am pretty sure that current mail clients assume that every MTA on the way supports  [https://tools.ietf.org/html/rfc6152](url). At least if I send a message with Thunderbird with default settings, it uses an 8bit-encoding (In fact, utf-8).

More to the point, passing UTF-8 encoded messages to `smtplib.send` yields the expected result (assuming the correct headers are set and only the body uses UTF-8, not the headers).

**Comment by d-maurer** [comment]
Viktor Dick wrote at 2020-7-14 01:49 -0700:
> ...
>More to the point, passing UTF-8 encoded messages to `smtplib.send` yields the expected result (assuming the correct headers are set and only the body uses UTF-8, not the headers).

Rereading your original post I have the impression that
the problem cause is `Products.MailHost`: because messages are at lower level
byte sequences,
`MailHost` should not convert a byte sequence into `str` - or at least
not guessing an encoding.

I would not put encoding guessing into `zope.sendmail` or sniff
encoding information from the message headers. Encoding handling
is implemented elsewhere (e.g. the `email` package); there is no need
to implement is another time.

**Comment by viktordick** [comment]
I agree. I posted the issue here because I could not exactly pinpoint the location in `Products.MailHost` where the issue would have to be fixed, while it was a two-liner if I place the fix in `zope.sendmail`. But the correct place is `Products.MailHost`.

Is it possible to move the issue there? Otherwise I will close this and re-open another issue there (and take another look at what needs to be fixed there).
